### PR TITLE
BR-2346 / Initiate Magento refunds on Balance

### PR DESCRIPTION
**Pull request by dhiraj-toptal**

### CODE REVIEW

**Comment by ynaxonBalance** [comment]
@dhiraj-toptal please add description of what is covered in this PR

File: Model/BalancepayMethod.php [hunk 1]
```diff
@@ -403,14 +403,19 @@ public function capture(InfoInterface $payment, $amount)
...
+            $response = $this->requestFactory
                 ->create(RequestFactory::CAPTURE_REQUEST_METHOD)
                 ->setPayment($payment)
                 ->setAmount($amount)
                 ->setBalanceVendorId($balanceVendorId)
                 ->process();
-        }
 
+            $charges = $response['charges'];
+            if (is_array($charges) && isset($charges[0])) {
```
Comment (ynaxonBalance) [inline Model/BalancepayMethod.php]: what is `$payment`? is it saved on the DB?
if so, is there a relation between `payment` and `invoice`?

File: Model/ChargedProcessor.php [hunk 2]
```diff
@@ -69,8 +76,19 @@ public function processChargedWebhook($params, $order)
             if (!$isBalancepayAuthCheckout) {
                 $orderPayment->capture(null);
             }
+            $orderPayment->setTransactionId($chargeId);
```
Comment (ynaxonBalance) [inline Model/ChargedProcessor.php]: shouldn't this be the Balance transaction id?

File: Model/Request/Refunds.php [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+            $responseFactory,
+            $accountManagement,
+            $region
+        );
+    }
+
+    /**
+     * @var string
+     */
+    protected $_topic;
```
Comment (ynaxonBalance) [inline Model/Request/Refunds.php]: why do we have underscore for this property?

File: Model/Response/Transactions/Capture.php [hunk 4]
```diff
@@ -18,6 +18,23 @@
  */
 class Capture extends AbstractResponse
 {
+    /**
+     * @var
+     */
+    private $_charges;
```
Comment (ynaxonBalance) [inline Model/Response/Transactions/Capture.php]: why underscore?

File: Observer/SalesOrderCreditmemoSaveAfter.php [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        $this->requestFactory = $requestFactory;
+        $this->balancepayRefundFactory = $balancepayRefundFactory;
+    }
+
+    public function execute(Observer $observer)
+    {
+        $creditMemo = $observer->getEvent()->getCreditmemo();
+        $creditMemoId = $creditMemo->getId();
+        $total = $creditMemo->getBaseGrandTotal();
+        $invoiceId = $creditMemo->getInvoiceId();
```
Comment (ynaxonBalance) [inline Observer/SalesOrderCreditmemoSaveAfter.php:65]: throw if no invoice id?

**Review by ynaxonBalance — COMMENTED** [review]
Looking good! Left some comments

File: Model/Request/Refunds.php [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+     * @return array
+     * @throws PaymentException
+     */
+    protected function getParams()
+    {
+        $params = [
+            'topic' => $this->_topic,
+            'amount' => $this->amount,
+            'chargeId' => $this->chargeId,
+            'reason' => 'other'
```
Comment (ynaxonBalance) [inline Model/Request/Refunds.php:215]: can't we get reason as input from credit memo?

File: Model/ChargedProcessor.php (hunk 2 above)
Comment (dhiraj-toptal) [reply Model/ChargedProcessor.php]: We will remove this as its not required to be saved on payment object.

File: Model/Request/Refunds.php (hunk 3 above)
Comment (dhiraj-toptal) [reply Model/Request/Refunds.php]: Removed.

File: Model/Response/Transactions/Capture.php (hunk 4 above)
Comment (dhiraj-toptal) [reply Model/Response/Transactions/Capture.php]: Removed.

File: Observer/SalesOrderCreditmemoSaveAfter.php (hunk 5 above)
Comment (dhiraj-toptal) [reply Observer/SalesOrderCreditmemoSaveAfter.php:65]: Didnt send the request to balance and added a notice message

File: Model/ChargedProcessor.php (hunk 2 above)
Comment (ynaxonBalance) [reply Model/ChargedProcessor.php]: removed?

File: Observer/SalesOrderCreditmemoSaveAfter.php [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+                $balancepayRefundModel = $this->balancepayRefundFactory->create();
+                $balancepayRefundModel->setData([
+                    'credit_memo_id' => $creditMemoId,
+                    'refund_id' => $refundId
+                ]);
+                $balancepayRefundModel->save();
+                $message = "Refund request has been sent to Balancepay.";
+            }
+        }
+        $this->messageManager->addNoticeMessage(__($message));
```
Comment (ynaxonBalance) [inline Observer/SalesOrderCreditmemoSaveAfter.php:89]: what is the message manager?

File: Observer/SalesOrderCreditmemoSaveAfter.php (hunk 5 above)
Comment (ynaxonBalance) [reply Observer/SalesOrderCreditmemoSaveAfter.php:65]: not sure I follow; what happens if there is no invoice id?

File: Model/ChargedProcessor.php (hunk 2 above)
Comment (dhiraj-toptal) [reply Model/ChargedProcessor.php]: Yes its removed, as we don't need transaction id saved, as Magento already saving it internally.

File: Observer/SalesOrderCreditmemoSaveAfter.php (hunk 5 above)
Comment (dhiraj-toptal) [reply Observer/SalesOrderCreditmemoSaveAfter.php:65]: We send a refund request to balance only if there's a Invoice-ID for a Credit-memo, cause we need charge-id to be sent to API.
So, if invoice_id is not available then we will create a credit memo in magento (but will not send refund request to balance), add on Magneto will add a notice message that refund request is not sent to Balance.

File: Observer/SalesOrderCreditmemoSaveAfter.php (hunk 7 above)
Comment (dhiraj-toptal) [reply Observer/SalesOrderCreditmemoSaveAfter.php:89]: Its a object to show success/error/notice message after page loads.

File: Observer/SalesOrderCreditmemoSaveAfter.php (hunk 5 above)
Comment (ynaxonBalance) [reply Observer/SalesOrderCreditmemoSaveAfter.php:65]: Can we fail the credit memo creation in such case?
If not, will the merchant be able to manually update the credit memo status?

File: Model/Request/Refunds.php (hunk 6 above)
Comment (ynaxonBalance) [reply Model/Request/Refunds.php:215]: @dhiraj-toptal Have you checked if we can get reason as input from user?

Comment (dhiraj-toptal) [reply Model/Request/Refunds.php:215]: We have a text input in the credit memo section, but if we need a specific value to be sent then we need a dropdown and that has to be customized.

File: Observer/SalesOrderCreditmemoSaveAfter.php (hunk 5 above)
Comment (dhiraj-toptal) [reply Observer/SalesOrderCreditmemoSaveAfter.php:65]: If we want to fail the credit memo then we need to check those conditions before credit-memo creation.
The invoice_id will be always be available though as we are creating credit memo from Invoice page, but we need to check if charge_id is available for that Invoice.
There's a canRefund class for Invoice where we can check if the charge_id and invoice_id in the balance_charges tabel.

Comment (dhiraj-toptal) [reply Observer/SalesOrderCreditmemoSaveAfter.php:65]: I am working on this to not show credit memo button on Invoice page whenever we dont have a charge related to invoice.

Comment (ynaxonBalance) [reply Observer/SalesOrderCreditmemoSaveAfter.php:65]: When we don't have charge id, let's cancel the credit memo for now. If possible, let's display a reason why it failed ("No charge id")

**Comment by dhiraj-toptal** [comment]
Updated PR

**Review by ynaxonBalance — APPROVED** [review]

**Pull request merged**